### ISSUE TITLE
Will the AIR SDK support Huawei HarmonyOS systems?

### ISSUE THREAD

**Issue by pengyangjin** [airsdk/Adobe-Runtime-Support#3327]
As Huawei's HarmonyOS becomes increasingly popular,it is very necessary for us to adapt this.This system requires the use of a file called hap which is similar to APK and IPA.
In China, the Huawei phone which equipped the HarmonyOS operating system is used by high-income individuals who are more willing to pay for our products, so we are very eager to support this operating system by AIR SDK.

**Comment by ajwfrost** [comment]
We're still struggling to get any coherent information about HarmanyOS and the direction it's going, how to develop for it, etc.

For example: this document is their "Creating an App" description:
https://developer.huawei.com/consumer/en/doc/app/agc-help-createapp-0000001146718717
and it says you can create an APK, or an RPK, or an APP. 

But that document seems to be about "AppGallery" .. maybe that's different?

If we just go to the basic site:
https://www.harmonyos.com/en/
it says you can develop apps with JavaScript or extended TypeScript, which means that AIR apps would not be possible. 
"Learn More" takes me to https://developer.huawei.com/consumer/cn/
But switching to English: https://developer.huawei.com/consumer/en/
is a completely different page and talks about "HMS Core" and then the AppGallery per the above.

Maybe "HMS Core" is what we need? It does mention a Java integration. But we still don't have any idea what the "HarmonyOS Next" is and whether this also supports Java-based applications, what their JNI support is like, etc.

Can you please provide a link to the information about the "hap" file you mention?

thanks

**Comment by RayGyoe** [comment]
> We're still struggling to get any coherent information about HarmanyOS and the direction it's going, how to develop for it, etc.
> 
> For example: this document is their "Creating an App" description: https://developer.huawei.com/consumer/en/doc/app/agc-help-createapp-0000001146718717 and it says you can create an APK, or an RPK, or an APP.
> 
> But that document seems to be about "AppGallery" .. maybe that's different?
> 
> If we just go to the basic site: https://www.harmonyos.com/en/ it says you can develop apps with JavaScript or extended TypeScript, which means that AIR apps would not be possible. "Learn More" takes me to https://developer.huawei.com/consumer/cn/ But switching to English: https://developer.huawei.com/consumer/en/ is a completely different page and talks about "HMS Core" and then the AppGallery per the above.
> 
> Maybe "HMS Core" is what we need? It does mention a Java integration. But we still don't have any idea what the "HarmonyOS Next" is and whether this also supports Java-based applications, what their JNI support is like, etc.
> 
> Can you please provide a link to the information about the "hap" file you mention?
> 
> thanks

The closest information should be NDK data:https://developer.huawei.com/consumer/cn/doc/harmonyos-guides-V5/ndk-development-overview-V5

**Comment by pengyangjin** [comment]
> Thanks - that actually looks quite promising from what little I can glean from it... shame there is no English variant, but I will ask our team member from Chengdu to have a look!

May I ask if there has been any progress on this issue? Are there any support plans in the future?

**Comment by AdamSun** [comment]
@ajwfrost Hi,any progress for the support of the Huawei's HarmonyOS?

**Comment by yishigulang** [comment]
My God, please support HarmonyOS!

**Comment by pengyangjin** [comment]
> My God, please support HarmonyOS!Very much needed!

the HarmonyOS users are very important for game developers in Chinese Mainland.They are wealthy and willing to spend money on games.

**Comment by ajwfrost** [comment]
With thanks for the link from @RayGyoe above, we were able to find some documentation in English for the 'NEXT' version of HarmonyOS, which is looking a bit more promising in terms of being able to still use native (C++) code for the AIR runtime:
https://developer.huawei.com/consumer/en/doc/harmonyos-guides-V5/application-dev-overview-V5

So we could try to look at a version of AIR built for this; it would be a lot of work to do this though. We had been looking at an alternative approach, starting discussions with a local company who may be able to help support this via a more web-centric approach, but it's early stages there.

I can see whether can we get a little further in a proof of concept of a 'traditional' AIR approach, using a native library and then having an ArkTS based application layer above this. Our earlier attempts ran into a bit of a brick wall with tools/documentation etc but from what I can see here, I will go back and ask someone to re-try using the latest info.

thanks

**Comment by Devin-Licastro** [comment]
Sounds promising, thanks for the update! This is the feature request that keeps us up at night :D (but I understand it's been super hard to get any info and you guys are doing your best)

**Comment by AdamSun** [comment]
@ajwfrost Hi,maybe u can contact the Huawei Technical Support Team.
[Technical Support](https://developer.huawei.com/consumer/en/support/contactUs/)

**Comment by ajwfrost** [comment]
@AdamSun yes we'd tried that, I didn't get a response when writing a message to them in English so I asked a colleague in our Chengdu office to do the same, she also didn't get anything back... anyway - I think we have a couple of potential routes at the moment that we can use to explore this further.

thanks

**Comment by pengyangjin** [comment]
> @AdamSun yes we'd tried that, I didn't get a response when writing a message to them in English so I asked a colleague in our Chengdu office to do the same, she also didn't get anything back... anyway - I think we have a couple of potential routes at the moment that we can use to explore this further.
> 
> thanks

Here is a possible contact link,you can have a try.
https://developer.huawei.com/consumer/cn/support/feedback/#/

**Comment by pengyangjin** [comment]
> [@AdamSun](https://github.com/AdamSun) yes we'd tried that, I didn't get a response when writing a message to them in English so I asked a colleague in our Chengdu office to do the same, she also didn't get anything back... anyway - I think we have a couple of potential routes at the moment that we can use to explore this further.
> 
> thanks

Is this project still progressing? Is there any latest news.Can you provide the contact information of the Chengdu team？

**Comment by pengyangjin** [comment]
@ajwfrost Is this project still progressing? Is there any latest news.Can you provide the contact information of the Chengdu team？

**Comment by wmtj** [comment]
In the past year, I have received many phone calls from Huawei HarmonyOS adaptation promotion specialist. Huawei hopes that our App can be adapted to HarmonyOS, and is willing to provide development guidance and support. I told them that as long as Adobe AIR is adapted to Hongmeng, our App will automatically adapt. However, Huawei's feedback is not aware that the Adobe AIR team has docking with them.

I don't know whether you have established effective docking with Huawei in the last six months. If not, I would provide their email to you so that you can contact them. @ajwfrost

**Comment by Devin-Licastro** [comment]
> 
> I don't know whether you have established effective docking with Huawei in the last six months. If not, I would provide their email to you so that you can contact them. [@ajwfrost](https://github.com/ajwfrost)

That sounds promising, we are getting more and more users upgrading and upset that our app no longer works, all we can suggest to them is an OS downgrade for now.

**Comment by ajwfrost** [comment]
@wmtj that would be very helpful, if you're able to introduce me to a mail thread we can try to get the clarifications and support to be able to move this one forwards a bit more. FYI the team we have in Chengdu aren't involved in the development of the AIR runtime, but we have a couple of people from there who are involved in our wider Adobe-related business practice. We're also talking with a third-party company who may be able to help with the wider ecosystem development for this... So, things are moving, albeit at a snail's pace..

thanks

**Comment by wmtj** [comment]
> [@wmtj](https://github.com/wmtj) that would be very helpful, if you're able to introduce me to a mail thread 
> 
I hava send you email ( @Harman.com ).

**Comment by Ender22** [comment]
Hey guys, any progress on this? Our CS team says they are getting more and more complaints every day of users no longer being able to use our app :(

**Comment by Ender22** [comment]
I'm still getting pushed on this daily, it would just be great to know if this is still being slowly worked on or not please

**Comment by AleDaDa** [comment]
Are there any  progress ?
